### PR TITLE
Fixing security vulnerabilities

### PR DESCRIPTION
**Pull request by dr-bizz**
### What I've done
I've fixed nearly most of the security vulnerabilities.

#### Updated the following packages:

- "karma": "6.3.12"
- "typescript": "3.9.6"
- "webpack": "4.44.2"
- "webpack-dev-server": "3.11.0"
- "jsencrypt": "3.1.0"
- "@types/fetch-mock": "7.3.5"
- "@typescript-eslint/parser": "5.10.1"
- "eslint": "7.32.0"
- "karma-typescript": "5.5.3"
- "karma-typescript-es6-transform": "5.5.3"
- "ts-loader": "8.3.0"

#### Removed the following packages

- "@types/es6-promise"
- "@types/whatwg-fetch"
- "typescript-eslint-parser"

#### Added the following new packages:

- "webpack-cli": "3.3.12"
- "@typescript-eslint/parser": "5.10.1"
- "karma-typescript-es6-transform": "5.5.3"

#### Altered some pages as needed with updating packages.

- Karma now uses chrome - PhantomJS had a lot of security vulnerabilities and is no longer maintained.
- Removed payment types - No longer needed as already defined.
- Updated Payment tests - Global `this` was no longer defined.
- Edited ESLint rules - Required for a newer version of TS 

------------------------
## Changes since previous version

- [x] Change from import cruPayments from 'cru-payments/dist/cru-payments' to import * as cruPayments from '@cruglobal/cru-payments/dist/cru-payments'

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -20,7 +21,7 @@
     "no-case-declarations": "error",
     "no-catch-shadow": "off",
     "no-class-assign": "error",
-    "no-cond-assign": "error",
+    "no-cond-assign": "warn",
```
Comment (wrandall22) [inline .eslintrc.json:24]: What necessitated this change?

File: src/payment-providers/ccp/ccp.spec.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import * as ccp from './ccp';
 
 import {once as mockOnce} from 'fetch-mock';
+var validKey: string = '';
```
Comment (wrandall22) [inline src/payment-providers/ccp/ccp.spec.ts:4]: Should this use `let` instead?

File: src/payment-providers/tsys/tsys.spec.ts [hunk 3]
```diff
@@ -9,6 +9,7 @@ interface TsysError {
   data: any;
 }
 /* eslint-enable no-undef */
+var validKey: string = '';
```
Comment (wrandall22) [inline src/payment-providers/tsys/tsys.spec.ts:12]: Should this use `let` instead?

File: karma.conf.js [hunk 4]
```diff
@@ -39,6 +39,7 @@ module.exports = function(config) {
     files: [
       'src/**/*.ts'
     ],
+    exclude: ["src/payment-providers/declarations.d.ts"],
```
Comment (wrandall22) [inline karma.conf.js:42]: Where is the other place this is defined?

File: src/payment-providers/declarations.d.ts [hunk 5]
```diff
@@ -9,58 +9,3 @@ declare module 'jsencrypt' {
     new (): JSEncryptPrototype;
   };
 }
-
-
-interface Symbol {
```
Comment (wrandall22) [inline src/payment-providers/declarations.d.ts:14]: Where are these defined elsewhere, or are they just unused?

Comment (dr-bizz) [reply src/payment-providers/declarations.d.ts:14]: They're defined in the new version of JSEncrypt.

File: karma.conf.js (hunk 4 above)
Comment (dr-bizz) [reply karma.conf.js:42]: There isn't another place it gets called. Testing works with it excluded.

For tests, Karma-typescript transpiles `.d.ts` files incorrectly, which results in failed tests as the Typescript compiler resolves them as ambient modules. Normally devs put their declaration files under a `@types` folder and exclude that folder from tests when using karma-typescript.

File: .eslintrc.json (hunk 1 above)
Comment (dr-bizz) [reply .eslintrc.json:24]: This was required for the luhn check, as this is the [formula](https://github.com/CruPCI/cru-payments/blob/master/src/credit-card/card-number/utils/luhn-check.ts#L10):
 `counter += (odd = !odd) ? digit : luhnArray[digit];`

The newer version of esLint throws an error at this code.

File: src/payment-providers/tsys/tsys.spec.ts (hunk 3 above)
Comment (dr-bizz) [reply src/payment-providers/tsys/tsys.spec.ts]: Fixed

File: src/payment-providers/ccp/ccp.spec.ts (hunk 2 above)
Comment (dr-bizz) [reply src/payment-providers/ccp/ccp.spec.ts]: Fixed

**Comment by wrandall22** [comment]
So, because of the code freeze for give site (which runs mid-October through early January) we won't deploy this until later.

**Comment by wrandall22** [comment]
Looks like this project uses the GitHub Releases feature for deploying releases to NPM (in conjunction with Travis in some way). At this point I think we can deploy a new version and test it out in stage by updating the version of cru-payments in give-web.

**Comment by wrandall22** [comment]
Unfortunately, it looks like we might need to switch to GHA in a separate PR first (I suppose we could manually deploy a version by running the commands Travis would otherwise run).

**Review by wrandall22 — APPROVED** [review]

**Comment by dr-bizz** [comment]
I think we should do a pre-release, making sure people know it's a pre-release by adding `beta` or `alpha` on the version.

**Comment by wrandall22** [comment]
Testing on give-web gets some errors:

```
TypeError: Cannot read properties of undefined (reading 'bankAccount')
    at BankAccountController.addCustomValidators (bankAccountForm.component.js:67:73)
    at bankAccountForm.component.js:57:12
    at Scope.$digest (angular.js:19270:1)
    at Scope.$apply (angular.js:19630:1)
    at angular.js:19934:1
    at TaskTracker.completeTask (angular.js:21403:1)
    at angular.js:6879:1
```

```
TypeError: Cannot read properties of undefined (reading 'creditCard')
    at CreditCardController (creditCardForm.component.js:42:33)
    at Object.invoke (angular.js:5208:1)
    at $controllerInit (angular.js:11814:1)
    at nodeLinkFn (angular.js:10627:1)
    at angular.js:11041:1
    at processQueue (angular.js:18075:1)
    at angular.js:18123:26
    at Scope.$digest (angular.js:19242:1)
    at Scope.$apply (angular.js:19630:1)
    at HTMLButtonElement.<anonymous> (angular.js:29127:1)
```

**Comment by wrandall22** [comment]
If we keep with putting this into `@cruglobal` we'll need to update the README to change the import statements.

**Comment by wrandall22** [comment]
Caleb helped me figure out the issue.

**Comment by dr-bizz** [comment]
@wrandall22 Was the issue related to `cru-payment` code or third-party?

**Comment by wrandall22** [comment]
I had to change from `import cruPayments from 'cru-payments/dist/cru-payments'` to `import * as cruPayments from '@cruglobal/cru-payments/dist/cru-payments'`

**Comment by dr-bizz** [comment]
That is noteworthy

**Comment by wrandall22** [comment]
Also, the update seems to have changed how jest is interacting with the library: `TypeError: Cannot redefine property: init`

**Comment by wrandall22** [comment]
I found some ways around the issue in the tests.

**Review by wrandall22 — APPROVED** [review]

**Pull request merged**